ecosystem/typescript/sdk_v2/src/transactions/transaction_builder/transaction_builder.ts

```diff
@@ -0,0 +1,512 @@
+import { getLedgerInfo } from "../../internal/general";
+import { getGasPriceEstimation } from "../../internal/transaction";
+import { HexInput, SigningScheme } from "../../types";
+import { NetworkToChainId } from "../../utils/api-endpoints";
+import { DEFAULT_MAX_GAS_AMOUNT, DEFAULT_TXN_EXP_SEC_FROM_NOW } from "../../utils/const";
+import {
+  AccountAuthenticator,
+  AccountAuthenticatorEd25519,
+  AccountAuthenticatorSecp256k1,
+} from "../authenticator/account";
+import {
+  TransactionAuthenticatorEd25519,
+  TransactionAuthenticatorFeePayer,
+  TransactionAuthenticatorMultiAgent,
+  TransactionAuthenticatorSecp256k1,
+} from "../authenticator/transaction";
+import {
+  ChainId,
+  EntryFunction,
+  FeePayerRawTransaction,
+  MultiAgentRawTransaction,
+  MultiSig,
+  MultiSigTransactionPayload,
+  RawTransaction,
+  Script,
+  TransactionPayloadEntryFunction,
+  TransactionPayloadMultisig,
+  TransactionPayloadScript,
+} from "../instances";
+import { SignedTransaction } from "../instances/signedTransaction";
+import {
+  GenerateTransactionOptions,
+  TransactionPayload,
+  AnyRawTransactionInstance,
+  GenerateTransactionPayloadData,
+  GenerateFeePayerRawTransactionArgs,
+  GenerateMultiAgentRawTransactionArgs,
+  GenerateRawTransactionArgs,
+  GenerateSingleSignerRawTransactionArgs,
+  SingleSignerTransaction,
+  AnyRawTransaction,
+  FeePayerTransaction,
+  MultiAgentTransaction,
+  EntryFunctionData,
+  MultiSigData,
+  ScriptData,
+  SimulateTransactionData,
+} from "../types";
+
+/**
+ * We are defining function signatures, each with its specific input and output.
+ * These are the possible function signature for our `generateTransactionPayload` function.
+ * When we call our `generateTransactionPayload` function with the relevant type properties,
+ * Typescript can infer the return type based on the appropriate function overload.
+ */
+export function generateTransactionPayload(args: EntryFunctionData): TransactionPayloadEntryFunction;
+export function generateTransactionPayload(args: ScriptData): TransactionPayloadScript;
+export function generateTransactionPayload(args: MultiSigData): TransactionPayloadMultisig;
+export function generateTransactionPayload(args: GenerateTransactionPayloadData): TransactionPayload;
+/**
+ * Builds a transaction payload based on the data argument and returns
+ * a transaction payload - TransactionPayloadScript | TransactionPayloadMultisig | TransactionPayloadEntryFunction
+ *
+ * @param args.data GenerateTransactionPayloadData
+ *
+ * @return TransactionPayload
+ */
+export function generateTransactionPayload(args: GenerateTransactionPayloadData): TransactionPayload {
+  // generate script payload
+  if ("bytecode" in args) {
+    const scriptPayload = new TransactionPayloadScript(
+      new Script(hexToBytes(args.bytecode), args.type_arguments, args.arguments),
+    );
+    return scriptPayload;
+  }
+
+  // generate multi sig payload
+  if ("multisigAddress" in args) {
```

Comment:
Guards like this are problematic if say these things are applied with multiple in the future, we may want to approach explicitly saying the expected type in the transaction payload data.